toolbox/relnotes/main.go

```diff
@@ -0,0 +1,784 @@
+	releaseTars   = flag.String("release-tars", "", "Directory of tars to sha256 sum for display")
+	repo          = flag.String("repo", "kubernetes", "Github repository")
+
+	// Global
+	branchHead = ""
+)
+
+// ReleaseInfo contains release related information to generate a release note.
+// NOTE: the prMap only includes PRs with "release-note" label.
+type ReleaseInfo struct {
+	startTag, releaseTag     string
+	prMap                    map[int]*github.Issue
+	releasePRs               []int
+	releaseActionRequiredPRs []int
+}
+
+func main() {
+	// Initialization
+	flag.Parse()
+	branchRange := flag.Arg(0)
+	startingTime := time.Now().Round(time.Second)
+
+	log.Printf("Boolean flags: full: %v, htmlize-md: %v, preview: %v, quiet: %v", *full, *htmlizeMD, *preview, *quiet)
+	log.Printf("Input branch range: %s", branchRange)
+
+	if *branch == "" {
+		// If branch isn't specified in flag, use current branch
+		var err error
+		*branch, err = u.GetCurrentBranch()
+		if err != nil {
+			log.Printf("not a git repository: %s", err)
+			os.Exit(1)
+		}
+	}
+
+	prFileName := fmt.Sprintf("/tmp/release-notes-%s-prnotes", *branch)
+	if *mdFileName == "" {
+		*mdFileName = fmt.Sprintf("/tmp/release-notes-%s.md", *branch)
+	}
+	log.Printf("Output markdown file path: %s", *mdFileName)
+	if *htmlFileName != "" {
+		log.Printf("Output HTML file path: %s", *htmlFileName)
+	}
+
+	if *githubToken == "" {
+		// If githubToken isn't specified in flag, use the GITHUB_TOKEN environment variable
+		*githubToken = os.Getenv("GITHUB_TOKEN")
+	} else {
+		token, err := u.ReadToken(*githubToken)
+		if err != nil {
+			log.Printf("failed to read Github token: %s", err)
+			os.Exit(1)
+		}
+		*githubToken = token
+	}
+	client := u.NewClient(*githubToken)
+
+	// End of initialization
+
+	// Gather release related information including startTag, releaseTag, prMap and releasePRs
+	releaseInfo, err := gatherReleaseInfo(client, branchRange)
+	if err != nil {
+		log.Printf("failed to gather release related information: %s", err)
+		os.Exit(1)
+	}
+
+	// Generating release note...
+	log.Printf("Generating release notes...")
+	err = gatherPRNotes(prFileName, releaseInfo)
+	if err != nil {
+		log.Printf("failed to gather PR notes: %s", err)
+		os.Exit(1)
+	}
+
+	// Start generating markdown file
+	log.Printf("Preparing layout...")
+	err = generateMDFile(client, releaseInfo.releaseTag, prFileName)
+	if err != nil {
+		log.Printf("failed to generate markdown file: %s", err)
+		os.Exit(1)
+	}
+
+	if *htmlizeMD && !u.IsVer(releaseInfo.releaseTag, verDotzero) {
+		// HTML-ize markdown file
+		// Make users and PRs linkable
+		// Also, expand anchors (needed for email announce())
+		projectGithubURL := fmt.Sprintf("https://github.com/%s/%s", *owner, *repo)
+		_, err = u.Shell("sed", "-i", "-e", "s,#\\([0-9]\\{5\\,\\}\\),[#\\1]("+projectGithubURL+"/pull/\\1),g",
+			"-e", "s,\\(#v[0-9]\\{3\\}-\\),"+projectGithubURL+"/blob/master/CHANGELOG.md\\1,g",
```

Comment:
Now `https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG.md` only includes links to the CHANGELOG of each release, i.e., CHANGELOG-1.x.md. So we shouldn't expand anchors to CHANGELOG.md any longer. Please use the file name of each release.

Reply:
Updated.